### DISCUSSION TITLE
MEGA65 support

### DISCUSSION THREAD

**Discussion by gardners** [cc65/cc65#1923] [Q&A]
We would love for CC65 to support the MEGA65, and make use of various features to help produce faster/more compact code for that target:

1. DMA memory copies/fills
2. Hardware 32x32 fixed point multiplier
3. Hardware 32/32 = 32.32 fixed point divider
4. Fast CPU mode
5. 4510 stack-relative opcodes
6. 32-bit compound instructions
7. Access to memory >64KB without bank switching
etc...

We're happy to help out, but would rather do a coordinated effort with yourselves, than do it in a way that doesn't fit with your architecture.

**Comment by mrdudz** [comment]
It's a huge undertaking. Several things need to be done:

first add assembler support:
- add support for new opcodes to the assembler (this seems to exist)
- provide linker configs for your target (missing)

then add support to the library:
- use extra capabilites of the hardware (new opcodes) in the runtime functions
- port the rest of the library (conio, drivers etc)

now add support to the simulator, so compiler changes can be tested in the regression tests
- update the cpu emulator
- add tests for this

last not least add support to the compiler:
- update the codegenerator
- improve the optimizer
- add a lot of regression tests to make sure you dont break something

that said, i wouldnt expect anyone doing this for you :=) especially when it comes to changes to the compiler itself, you are very much on your own (as there arent even many ppl around who understand the code).

you may want to start sending pull requests adding things in roughly this order. make sure to use the same codestyle as in the rest of the code. and make sure to keep the documentation in sync.

adding anything related to address space > 64k will likely require even more effort (this should probably be a seperated effort, and include adding 65816 too). i dont expect this to happen anytime soon though.

**Comment by gardners** [comment]
Hello,

Thanks for the clear directions.  This is the information that I was lacking.  We're not expecting yourselves to do it, but wanted to make sure that we come at things from the right angle.

The linker config side is a bit interesting, as running from C64 mode is probably still the most sensible thing to do on a C65, as the many different versions of ROMs mean that some things move around the place in C65 mode.  Also, unlike a C128, all enhanced features are available from C64 mode. Thus I'm tempted to work first on a target of "MEGA65 in C64 mode".  As you say, its a fairly big body of work, so we'll likely think about things for a while, and dive in when we can dedicate the time to it -- or find someone from the MEGA65 community who is keen to get their hands dirty.

**Comment by oliverschmidt** [comment]
I fully agree to everything pointed out by @mrdudz (thanks!).

I'd like to add another aspect: As already mentioned by @mrdudz there's not much knowledge about the compiler internals in the cc65 community. So if someone from the MEGA65 community contributes large changes to the compiler which seem to further complicate the compiler in a significant way I might be very reluctant to merge them - even if they are "correct" and even if there are a lot of regression tests.

**Comment by gardners** [comment]
I'm anticipating that in the first instance we would only be looking to modify the code snippets that the code generator produces. That would yield considerable low-hanging fruit.  

Very long term, taking advantage of the 16-bit stack would be interesting to explore, but would require considerably more work, as it would require merging the processor and CC65 stacks back into a single stack.  The main benefit would be greatly improving the speed of function calls, and functions that access stack-based arguments.  This would also be of direct benefit to the 65CE02 target.  Not sure if the 65816 has the same extensions or not.  Anyway, we'll see when we get to it.
